lsp/src/text_document.ml

```diff
@@ -159,3 +164,59 @@ let apply_content_changes ?version t changes =
     | Some version -> { document with version }
   in
   { t with document }
+
+type change =
+  { start : int
+  ; stop : int
+  ; replacement : string
+  }
+
+module Edit_map = Map.Make (struct
+  type t = int
+
+  let compare = Int.compare
+end)
+
+let add_edit map change =
+  (* TODO check non overlapping property *)
+  Edit_map.update map ~key:change.start ~f:(function
+      | None -> Some [ change ]
+      | Some changes -> Some (change :: changes))
+
+let rec simplify_changes acc text find_position (changes : TextEdit.t list) =
+  match changes with
+  | [] -> acc
+  | { range; newText = replacement } :: changes ->
+    let start, stop = find_position ~utf8:text range in
+    let change = { start; stop; replacement } in
+    let acc = add_edit acc change in
+    simplify_changes acc text find_position changes
+
+let apply_changes text encoding changes =
+  let find_position =
+    match encoding with
+    | `UTF8 -> find_offset_8
+    | `UTF16 -> find_offset_16
+  in
+  let simplified = simplify_changes Edit_map.empty text find_position changes in
+  let pos = ref 0 in
```

Comment:
I think this keeps an index into `text`? (a comment would be helpful) 

I just don't understand the comment "if this is an insert, it's allowed to increase the position" - if `pos` is an index in `text`, shouldn't it remain unchanged on an insertion?

Also, in `pos := max !pos stop` - can `stop` be ever less than `pos` ?

Reply:
Even though I wrote this code only 2 days ago, I'm already not particularly confident in the implementation.  Feel free to improve it and add tests.

>  if pos is an index in text, shouldn't it remain unchanged on an insertion?

If it's an insert, then we have `start = stop`, which means that `pos := max !pos stop` will not advance past the start position of the edit. This should answer your question.

> Also, in pos := max !pos stop - can stop be ever less than pos ?

Only if we provide an invalid range where `start > end_`